libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -63,28 +60,54 @@ public MicrosoftAppCredentials(string appId, string password)
         public virtual string OAuthEndpoint { get { return AuthenticationConstants.ToChannelFromBotLoginUrl; } }
         public virtual string OAuthScope { get { return AuthenticationConstants.ToChannelFromBotOAuthScope; } }
 
-        protected readonly string TokenCacheKey;
 
         /// <summary>
-        /// Adds the host of service url to <see cref="MicrosoftAppCredentials"/> trusted hosts.
+        /// TimeWindow which controlls how often the token will be automatically updated
         /// </summary>
-        /// <param name="serviceUrl">The service URL.</param>
-        /// <remarks>If expiration time is not provided, the expiration time will DateTime.UtcNow.AddDays(1).</remarks>
-        public static void TrustServiceUrl(string serviceUrl)
-        {
-            TrustServiceUrl(serviceUrl, DateTime.UtcNow.Add(TimeSpan.FromDays(1)));
-        }
+        public static TimeSpan AutoTokenRefreshTimeSpan { get; set; } = TimeSpan.FromMinutes(10);
```

Comment:
Is there a concurrency concern here? TimeStap is a ValueType, and I don't remember what the rules are for those in static methods. I *suspect* there is a potential race here where corrupted data could be read.

Reply:
My understanding is that global initialization happens as part of assembly loading, and so there should not be a concurrency issue here.

---
In reply to: [198675926](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198675926) [](ancestors = 198675926)